pkg/controller/factory/base_controller.go

```diff
@@ -0,0 +1,158 @@
+		// this will be recovered using HandleCrash()
+		panic("timeout waiting for informer cache")
+	}
+	klog.V(5).Infof("Caches synced for controller %s", c.name)
+
+	var workerWaitGroup sync.WaitGroup
+
+	// queueContext is used to track and initiate queue shutdown
+	queueContext, queueShutdownComplete := context.WithCancel(context.TODO())
+
+	for i := 1; i <= workers; i++ {
+		klog.Infof("Starting #%d worker of %s controller ...", i, c.name)
+		workerWaitGroup.Add(1)
+		go func() {
+			defer func() {
+				klog.Infof("Shutting down worker of %s controller ...", c.name)
+				workerWaitGroup.Done()
+			}()
+			c.runWorker(queueContext)
+		}()
+	}
+
+	// runPeriodicalResync is independent from queue:w
+	go c.runPeriodicalResync(ctx, c.resyncEvery)
+
+	// Handle controller shutdown
+
+	<-ctx.Done() // wait for controller context to be cancelled
+	klog.Infof("Shutting down %s ...", c.name)
+
+	c.syncContext.Queue().ShutDown() // shutdown the controller queue first
+	for !c.syncContext.Queue().ShuttingDown() {
+		time.Sleep(10 * time.Millisecond) // don't race too much with the queue
+	}
+	queueShutdownComplete() // signal workers the queue is terminated
+
+	// wait for all workers to finish their jobs
+	workerWaitGroup.Wait()
+
+	shutdownComplete() // cancel the shutdown context
+}
+
+func (c *baseController) Sync(ctx context.Context, syncCtx SyncContext) error {
+	return c.sync(ctx, syncCtx)
+}
+
+func (c *baseController) ShutdownContext() context.Context {
+	return c.shutdownContext
+}
+
+func (c *baseController) runPeriodicalResync(ctx context.Context, interval time.Duration) {
+	if interval == 0 {
+		return
+	}
+	go wait.UntilWithContext(ctx, func(ctx context.Context) {
+		if err := c.sync(ctx, c.syncContext.(syncContext).withRuntimeObject(nil)); err != nil {
+			utilruntime.HandleError(fmt.Errorf("periodical resync of controller %s failed: %v", c.name, err))
+		}
+	}, interval)
+}
+
+// runWorker runs a single worker
+// The worker is asked to terminate when the passed context is cancelled and is given terminationGraceDuration time
+// to complete its shutdown.
+func (c *baseController) runWorker(queueCtx context.Context) {
+	// workerContext is used to signal when the processing of the current item is done and the queue is shutdown
+	workerContext, workerDone := context.WithCancel(context.Background())
+
+	go func() {
+		defer workerDone()
+		for {
+			select {
+			case <-queueCtx.Done():
+				return
+			default:
+				c.processNextWorkItem(queueCtx)
+			}
+		}
+	}()
+
+	// wait for the queue to be marked as shutdown
+	<-queueCtx.Done()
+
+	// sets a counter for the processor to finish its job.
+	// by using the workerContext as parent context either the workerContext is complete or terminationGraceDuration pass (whatever comes first).
+	// this allows the worker to gracefully terminate its job.
+	workerDeadline, workerDeadlineCancel := context.WithTimeout(workerContext, c.terminationGraceDuration)
```

Comment:
what use case is this trying to solve?
if `queueCtx` is done, the controller is shutting down, you need to wait for the workers anyways before you can release leader election. You can choose the timeout at the top level if you want using select but hidden leak is not what I'd expect

Reply:
what you mean with hidden leak? I thought the comment made this clear... each worker, independently gets its own 10 seconds to finish.

Reply:
But you don't kill it, you leave it running in the background - canceling context doesn't mean the code has ended, only that it is being signaled to cancel when reaching an operation checking the signal (ctx). It may as well be running for a bit. As a user, when `Run` finishes I expect all business logic to be safely off.

Is this meant as a fail safe so program graceful termination doesn't stuck? I think I'd rather keep it around the leader election code when we'd wait for graceful termination of controllers and set a timer for deadline forcefully exiting the program.

Reply:
Figured this out via phone. I'm going to remove the termination grace period and we are going to handle the graceful timeout from upper level, as the only way to guarantee the go routine is force killed is SIGKILL.

@tnozicka is going to open a PR later that is adding ControllerManager which will handle this.